certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => frontBefore == popped, "previous front is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front is previous at(1) after calling popFront                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextFront = at_(1);
+    popFront@withrevert();
+
+    assert !lastReverted <=> front() == nextFront, "front is adjusted";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(x) is preserved and offset to at(x - 1) after calling popFront                                             |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontConsistency {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 before = at_(key);
+
+    popFront@withrevert();
+    bool success = !lastReverted;
+    
+    assert !lastReverted <=> before == at_(key - 1), "values are offset and not modified";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popBack removes an element from the end of the queue                                          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popBack {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    bool emptyBefore = empty();
+    bytes32 backBefore = back@withrevert();
+
+    bytes32 popped = popBack@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => backBefore == popped, "previous back is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: back is previous at(length - 2) after calling popBack                                                         │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popBackExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextBack = at_(length() - 2);
+    popBack@withrevert();
+
+    assert !lastReverted => back() == nextBack, "back is adjusted";
+}
```

Comment:
This is a consequence of `popBackConsistency` + `queueEndings`. I'd remove that rule.